pkg/pool-stable-phantom/test/StablePhantomPool.test.ts

```diff
@@ -727,6 +731,96 @@ describe('StablePhantomPool', () => {
+            await pool.enableRecoveryMode(admin);
+          });
+
+          itJoinsExactBPTOutCorrectly();
+        });
+
+        function itJoinsExactBPTOutCorrectly() {
+          it('fails if not initialized', async () => {
+            await expect(pool.joinGivenOut({ bptOut: fp(2), token })).to.be.revertedWith('UNINITIALIZED');
+          });
+
+          context('once initialized', () => {
+            sharedBeforeEach('initialize pool', async () => {
+              await pool.init({ recipient, initialBalances });
+            });
+
+            it('grants exact BPT for token in', async () => {
+              const previousBptBalance = await pool.balanceOf(recipient);
+              // 20% of previous balance
+              const bptOut = pct(previousBptBalance, 0.2);
+              const expectedAmountIn = await pool.estimateTokenInGivenBptOut(token, bptOut);
+
+              const result = await pool.joinGivenOut({ from: recipient, recipient, bptOut, token });
+
+              // Only token in should be the one transferred
+              expect(result.amountsIn[tokenIndex]).to.be.equalWithError(expectedAmountIn, 0.001);
+              expect(result.amountsIn.filter((_, i) => i != tokenIndex)).to.be.zeros;
+
+              // Make sure received BPT is close to what we expect
+              const currentBptBalance = await pool.balanceOf(recipient);
+              expect(currentBptBalance.sub(previousBptBalance)).to.be.equal(bptOut);
+            });
+
+            it('can tell how many tokens it will receive', async () => {
+              const previousBptBalance = await pool.balanceOf(recipient);
+              // 20% of previous balance
+              const bptOut = pct(previousBptBalance, 0.2);
+
+              const queryResult = await pool.queryJoinGivenOut({ recipient, bptOut, token });
+
+              expect(queryResult.bptOut).to.be.equal(bptOut);
+              expect(queryResult.amountsIn.filter((_, i) => i != tokenIndex)).to.be.zeros;
+
+              const result = await pool.joinGivenOut({ from: recipient, bptOut, token });
+              // Query and join should match exactly
+              expect(result.amountsIn[tokenIndex]).to.equal(queryResult.amountsIn[tokenIndex]);
+            });
+
+            it('join and joinSwap give the same result', async () => {
+              const previousBptBalance = await pool.balanceOf(recipient);
+              // 20% of previous balance
+              const bptOut = pct(previousBptBalance, 0.325);
+
+              const queryResult = await pool.queryJoinGivenOut({ recipient, bptOut, token });
+
+              const amountIn = await pool.querySwapGivenOut({
+                from: recipient,
+                in: token,
+                out: pool.bpt,
+                amount: bptOut,
+                recipient: lp,
+              });
+
+              expect(amountIn).to.be.equalWithError(queryResult.amountsIn[tokenIndex], 0.00001);
```

Comment:
Does this test fail with strict equality?

Reply:
Ah must be because of the swap fees. Can you test locally that if the swap fees are zero these two match exactly (just a temporary local change, don't add it to the suite)? If that happens we can then merge.

Reply:
They do indeed match.